src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -299,33 +334,208 @@ private BoundExpression RewriteStringConcatenationOneExpr(BoundExpression lowere
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__op_Implicit_ToReadOnlySpanOfChar, out MethodSymbol stringImplicitConversionToReadOnlySpan, isOptional: true) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_ReadOnlySpan_T__ctor_Reference, out MethodSymbol readOnlySpanWrappingCtorGeneric, isOptional: true))
+            {
+                // One of args is a char or a ReadOnlySpan<char> (e.g. from previous rewrite). We can use span-based concatenation, which takes a bit more IL, but avoids allocation of a string from ToString call in case of char.
+                // And since implicit conversion from string to span is a JIT intrinsic, the resulting code ends up being faster than the one generated from the "naive" case with ToString call
+                var readOnlySpanOfChar = readOnlySpanWrappingCtorGeneric.ContainingType.Construct(_compilation.GetSpecialType(SpecialType.System_Char));
+                var readOnlySpanCtorRefParamChar = readOnlySpanWrappingCtorGeneric.AsMember(readOnlySpanOfChar);
+
+                var locals = ArrayBuilder<LocalSymbol>.GetInstance(capacity: previousLocals.Count + 1);
+                locals.AddRange(previousLocals);
+
+                var wrappedFirst = WrapIntoReadOnlySpanOfCharIfNecessary(loweredFirst, stringImplicitConversionToReadOnlySpan, readOnlySpanCtorRefParamChar, out var firstTempLocal);
+                var wrappedSecond = WrapIntoReadOnlySpanOfCharIfNecessary(loweredSecond, stringImplicitConversionToReadOnlySpan, readOnlySpanCtorRefParamChar, out var secondTempLocal);
+                var wrappedThird = WrapIntoReadOnlySpanOfCharIfNecessary(loweredThird, stringImplicitConversionToReadOnlySpan, readOnlySpanCtorRefParamChar, out var thirdTempLocal);
+                var wrappedFourth = WrapIntoReadOnlySpanOfCharIfNecessary(loweredFourth, stringImplicitConversionToReadOnlySpan, readOnlySpanCtorRefParamChar, out var fourthTempLocal);
+
+                locals.AddIfNotNull(firstTempLocal);
+                locals.AddIfNotNull(secondTempLocal);
+                locals.AddIfNotNull(thirdTempLocal);
+                locals.AddIfNotNull(fourthTempLocal);
+
+                var concatCall = BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, stringConcatSpans, ImmutableArray.Create(wrappedFirst, wrappedSecond, wrappedThird, wrappedFourth));
+
+                return new BoundSequence(
+                    syntax,
+                    locals.ToImmutableAndFree(),
+                    ImmutableArray<BoundExpression>.Empty,
+                    concatCall,
+                    concatCall.Type);
+            }
+
+            Debug.Assert(!firstType.IsReadOnlySpanChar() && !secondType.IsReadOnlySpanChar() && !thirdType.IsReadOnlySpanChar() && !fourthType.IsReadOnlySpanChar());
+            Debug.Assert(previousLocals.Count == 0);
```

Comment:
>Debug.Assert(previousLocals.Count == 0); [](http://example.com/codeflow?start=12&length=40)

This assumption isn't obvious to me at the moment.

Reply:
> This assumption isn't obvious to me at the moment.

Now it looks just wrong. This assert fails for the following unit-test, and, if all failing asserts are disabled, compiler crashes in stack optimizer:
```

        [Fact]
        public void Test3()
        {
            var source = """
                public class C 
                { 
                    static void Main()
                    {
                        var a = "a";
                        var b = 'b';
                        var c = "c";
                        var d = 'd';
                
                        System.Console.WriteLine((a + b) + (c + d)); 
                    }
                }
                """;

            var comp = CreateCompilation(source, options: TestOptions.ReleaseExe, targetFramework: TargetFramework.Net80);
            comp.MakeMemberMissing(WellKnownMember.System_String__Concat_ReadOnlySpanReadOnlySpanReadOnlySpanReadOnlySpan);

            var verifier = CompileAndVerify(compilation: comp, expectedOutput: RuntimeUtilities.IsCoreClr8OrHigherRuntime ? "abcd" : null, verify: /*ExecutionConditionUtil.IsCoreClr ? default : */Verification.Skipped).VerifyDiagnostics();

 //           verifier.VerifyIL("C.Main", "");
        }
```